Data/HashMap/Base.hs

```diff
@@ -431,22 +456,116 @@ member k m = case lookup k m of
 -- | /O(log n)/ Return the value to which the specified key is mapped,
 -- or 'Nothing' if this map contains no mapping for the key.
 lookup :: (Eq k, Hashable k) => k -> HashMap k v -> Maybe v
-lookup k0 m0 = go h0 k0 0 m0
+#if __GLASGOW_HASKELL__ >= 802
+-- GHC does not yet perform a worker-wrapper transformation on
+-- unboxed sums automatically. That seems likely to happen at some
+-- point (possibly as early as GHC 8.6) but for now we do it manually.
+lookup k m = case lookup# k m of
+  (# (# #) | #) -> Nothing
+  (# | a #) -> Just a
+{-# INLINE lookup #-}
+
+lookup# :: (Eq k, Hashable k) => k -> HashMap k v -> (# (# #) | v #)
+lookup# k m = lookupCont (\_ -> (# (# #) | #)) (\v _i -> (# | v #)) (hash k) k m
+{-# INLINABLE lookup# #-}
+#else
+lookup k m = lookup' (hash k) m
+{-# INLINABLE lookup #-}
+#endif
+
+-- | lookup' is a version of lookup that takes the hash separately.
+-- It is used to implement alterF.
+lookup' :: Eq k => Hash -> k -> HashMap k v -> Maybe v
+#if __GLASGOW_HASKELL__ >= 802
+-- GHC does not yet perform a worker-wrapper transformation on
+-- unboxed sums automatically. That seems likely to happen at some
+-- point (possibly as early as GHC 8.6) but for now we do it manually.
+lookup' h k m = case lookup'# h k m of
+  (# (# #) | #) -> Nothing
+  (# | a #) -> Just a
+{-# INLINE lookup' #-}
+
+lookup'# :: Eq k => Hash -> k -> HashMap k v -> (# (# #) | v #)
+lookup'# h k m = lookupCont (\_ -> (# (# #) | #)) (\v _i -> (# | v #)) h k m
+{-# INLINABLE lookup'# #-}
```

Comment:
For my own edification, why are some things INLINE and others INLINABLE? Is there any rule to follow when deciding which one?

Reply:
I'm not the best at that. `lookup'#` is in the nature of a "worker". Once `lookupCont` inlines into it, it will actually be fairly large. We therefore generally *don't* want to actually inline it at call sites. But we *do* want to make sure that GHC gives it an unfolding in the interface (`.hi`) file. That will allow the specializer to specialize it (in other modules) to the key types it's actually used with.

We follow GHC's own practice (when it performs the worker-wrapper transformation on *products*) of marking the wrappers `INLINE`. We really really want them to inline in most cases to prevent "boxing" (in this case the allocation of `Just` constructors). In *most* cases, inlining the wrapper will very quickly expose opportunities for simplification, and very likely *reduce* code size.